Add IBL high pass spatial filter to preprocessing

JoeZiminski opened the pull request:
PR to add the [IBL high-pass spatial filter](https://github.com/int-brain-lab/ibl-neuropixel/blob/15f85d7d3ec47422f1b5c64bdd47d9d6b1459a09/src/neurodsp/voltage.py#L154) into SI preprocesing (from IBL version 0.4.0 2022-10-28).

Thanks for taking a look, some questions I had:

* The original IBL function num_channels x num_samples, so here [transpose are performed](https://github.com/JoeZiminski/spikeinterface/blob/de7ee4e1508aa846a18f55e9d8f9d756b0f2a308/spikeinterface/preprocessing/highpass_spatial_filter.py#L125) in get_traces()  (for last IBL PR I just swapped the dims in the original function, but this function is more complex with sub functions, so would avoid if possible here, see [this comment](https://github.com/SpikeInterface/spikeinterface/issues/1025#issuecomment-1303176754_))

* This function (and the bad channel detection) require quite a lot of IBL functions. Is it worth vendoring these in a specific folder, or splitting IBL functions per-use case as is currently done in this PR?

* The tests at the moment are a bit excessive, just testing all possible combinations as a sanity check. These can be split up a bit, but I wasn’t quite sure of how to use the syntax for the SI organisation of tests. Is it possible to add a pytest.mark.parameterize for test functions outside of the pytest class?

* The IBL function has a [section for processing on sub-selection of channels](https://github.com/int-brain-lab/ibl-neuropixel/blob/15f85d7d3ec47422f1b5c64bdd47d9d6b1459a09/src/neurodsp/voltage.py#L174). For now I removed this as was not sure how it would play with the channel_indicies option in get traces (in fact, a general question I am a little unsure how to use this channel_indicies argument).

* In the IBL pipeline [automatic gain control](https://github.com/JoeZiminski/spikeinterface/blob/de7ee4e1508aa846a18f55e9d8f9d756b0f2a308/spikeinterface/preprocessing/highpass_spatial_filter.py#L197) is applied by convolving each trace with a Hanning window, to reduce difference between channel magnitudes before filtering. The default window size 300 samples in the IBL (who use NP, about 100 ms with NP sampling rate of 30000 Hz). I don’t see why this window size would need to change across recording lengths / probes, but I thought worth raising just in case.

alejoe91 commented on spikeinterface/preprocessing/highpass_spatial_filter.py:
```diff
@@ -0,0 +1,280 @@
+                     n_channel_taper will be set to the number of
+                     padded channels. Otherwise, the passed value will be used.
+
+    agc_options: Options for automatic gain control. By default, gain control
```
@JoeZiminski What does gain control do exactly? Can you explain a bit more here? From the implementation, it seems to be some sort of scaling, but I wonder whether this should be a separate step, decoupledfrom the actual spatial filter. What do you and @oliche think?

alejoe91 replied:
Just saw you explained it in the PR text. Fine to leave it here :)

alejoe91 commented on spikeinterface/preprocessing/highpass_spatial_filter.py:
```diff
@@ -0,0 +1,280 @@
+    def __init__(self, recording, n_channel_pad=None, n_channel_taper=None, agc_options="default", butter_kwargs=None):
+        BasePreprocessor.__init__(self, recording)
+
+        n_channel_pad, agc_options, butter_kwargs = self.handle_args(recording,
```
We should check another important aspect: spatial kfilt assumes that channels are sorted by depth, but this is not always the case, so we should add a check and raise an error / warning if they are not sorted

alejoe91 reviewed (commented):
Thanks @JoeZiminski 

Left a couple of comments for discussion :)

alejoe91 commented:
> PR to add the [IBL high-pass spatial filter](https://github.com/int-brain-lab/ibl-neuropixel/blob/15f85d7d3ec47422f1b5c64bdd47d9d6b1459a09/src/neurodsp/voltage.py#L154) into SI preprocesing (from IBL version 0.4.0 2022-10-28).
> 
> Thanks for taking a look, some questions I had:
> 

Thanks @JoeZiminski!

> * The original IBL function num_channels x num_samples, so here [transpose are performed](https://github.com/JoeZiminski/spikeinterface/blob/de7ee4e1508aa846a18f55e9d8f9d756b0f2a308/spikeinterface/preprocessing/highpass_spatial_filter.py#L125) in get_traces()  (for last IBL PR I just swapped the dims in the original function, but this function is more complex with sub functions, so would avoid if possible here, see [this comment](https://github.com/SpikeInterface/spikeinterface/issues/1025#issuecomment-1303176754_))

Fine by me!

> * This function (and the bad channel detection) require quite a lot of IBL functions. Is it worth vendoring these in a specific folder, or splitting IBL functions per-use case as is currently done in this PR?

Maybe then easier to have an `ibl_utils.py` file in `preprocessing`

> * The tests at the moment are a bit excessive, just testing all possible combinations as a sanity check. These can be split up a bit, but I wasn’t quite sure of how to use the syntax for the SI organisation of tests. Is it possible to add a pytest.mark.parameterize for test functions outside of the pytest class?

That's ok!

> * The IBL function has a [section for processing on sub-selection of channels](https://github.com/int-brain-lab/ibl-neuropixel/blob/15f85d7d3ec47422f1b5c64bdd47d9d6b1459a09/src/neurodsp/voltage.py#L174). For now I removed this as was not sure how it would play with the channel_indicies option in get traces (in fact, a general question I am a little unsure how to use this channel_indicies argument).

So this can be tricky: the `channel_indices` in the `get_traces` of a `BaseSegment` object can either be an array of indices or a slice, so both cases should be handled. So basically `channel_indices` is exactly the same as `collection` in the IBL implementation. This parameter should not change the processing, but only allow to retrieve a sub-selection of the channels on the fly (e.g. for plotting). To do that, you can simply update [this line](https://github.com/JoeZiminski/spikeinterface/blob/add_ibl_high_pass_spatial_filter/spikeinterface/preprocessing/highpass_spatial_filter.py#L131) to `return traces[:, channel_indices]`


> * In the IBL pipeline [automatic gain control](https://github.com/JoeZiminski/spikeinterface/blob/de7ee4e1508aa846a18f55e9d8f9d756b0f2a308/spikeinterface/preprocessing/highpass_spatial_filter.py#L197) is applied by convolving each trace with a Hanning window, to reduce difference between channel magnitudes before filtering. The default window size 300 samples in the IBL (who use NP, about 100 ms with NP sampling rate of 30000 Hz). I don’t see why this window size would need to change across recording lengths / probes, but I thought worth raising just in case.

I wonder whether this step should be a separate one now, since it does process the signal in a non-trivial way and one could also use it before other kind of preprocessing. What do you and @oliche think?

JoeZiminski commented:
Thanks a lot for your feedback Alessio, also I made some updates below as discussed with Samuel. Regarding the gain control, I see in the IBL pipeline it is applied in a number of cases (e.g. during common average referencing), and so it may be worth splitting out. Olivier would know best though.

A few changes / things to discuss: 

The dims are now swapped across all functions for consistency. 

The cosine taper is the same implementation as SI, however with extra function to select bounds, so the SI call in get_chunks_with_margin() could call this will full bounds to unify the cases, as ```taper_ibl = fcn_cosine([0, margin])(np.arange(margin))```.

Added a couple more  test cases and fixed the DEBUG implementation.

I swapped out the scipy.signal.fftconvolve for IBL fftconvolve, they are very similar but are slightly different in a few cases. In some instances when the window is small, or close to the number of samples in size, they are slightly different. For now I just avoided these cases in the test, but this probably isn’t the best approach. Happy to revert back to IBL fftconvolve for consistency. 

An interesting consideration is how to handle the filter across probes of different channel number / lengths, and we discussed having some meaningful units for the Wn, the % Nyquist cutoff on the high-pass filter used to filter across the channels. It would be nice to specify in some units that would be meaningful across probes with different number of channels. This could be expressed in units of distance e.g. cycles/meter, however this might be a little unintuitive. It might also make sense to specify as cycles/probe length [1].

That’s a good point about the channel ordering. What would be the best functions to call to retrieve the correct channel ordering / test the channel order? I realised I do not understand something – for the spatial channel ordering the 1-shank case is clear, but how are the 4-shank channels ordered in 1D in a spatially contagious way that makes sense during filtering?


[1] I believe taking num_channels / probe is equivalent to num_samples / second, i.e. increasing the number of channels on the probe is analogous to increasing the sampling rate. Setting probe = 1 is analogous to seconds =  1 so everything is easy to interpret. 
For a 384 channel probe the Nyquist frequency would be 192 cycles, the IBL default Wn=0.01, we can have 192*0.01 = 1.92 cycles/probe as our high-pass filter cutoff. It is then easier to align across probes, e.g. for a 32-channel probe the default would filter 16*0.01 = 0.16 cycles/probe, and so the cutoff may need adjusting  for equivalency (though am finding this is a little hard to conceptualise and would need to test).

alejoe91 commented:
@JoeZiminski I pushed some changes that:

- automatically sort and sort back traces in case the locations are not sorted
- general imporvements of naming, docstrings, and checks
- changed `agc_option` from "default" to "ibl"

Somehow the `test_add_trend_and_check_deleted` is failing. Could you double check on your side? Everything else passes :)

samuelgarcia commented on spikeinterface/preprocessing/phase_shift.py:
```diff
@@ -82,10 +82,12 @@ def get_traces(self, start_frame, end_frame, channel_indices):
         # this return a copy with margin  + taper on border always
         traces_chunk, left_margin, right_margin = get_chunk_with_margin(self.parent_recording_segment,
-                                                                        start_frame, end_frame, channel_indices,
+                                                                        start_frame, end_frame, slice(None),
```
I think we should keep channel_indices. To avoid loading and preprocess to much in the upstream node.

samuelgarcia replied:
But then we have to take care of this
traces_shift = apply_fshift_sam(traces_chunk, self.sample_shifts[channel_indices], axis=0)

alejoe91 replied:
ok, but we need one of the 2

samuelgarcia commented on spikeinterface/preprocessing/tests/test_highpass_spatial_filter.py:
```diff
@@ -0,0 +1,190 @@
+from spikeinterface.core.testing_tools import generate_recording
+import spikeinterface.widgets as sw
+
+try:
+    import spikeglx
+    import neurodsp.voltage as voltage
+    HAVE_IBL_NPIX = True
+except ImportError:
+    HAVE_IBL_NPIX = False
```
Testing against IBL should be done locally only and not on github.
I propose to make local import in the function

alejoe91 replied:
yep, removing

samuelgarcia commented on spikeinterface/preprocessing/tests/test_highpass_spatial_filter.py:
```diff
@@ -0,0 +1,190 @@
+    options = dict(lagc=lagc, ntr_pad=25, ntr_tap=50, butter_kwargs=None)
+    print(options)
+
+    ibl_data, si_recording = get_ibl_si_data()
```
I would not do this on gihut actions. It will slow down even more.
We can keep this test locally but skip if on github.

samuelgarcia commented on spikeinterface/preprocessing/highpass_spatial_filter.py:
```diff
@@ -0,0 +1,290 @@
+            order_f = None
+            order_r = None
+        else:
+            order_f = np.argsort(locs_mono)
```
Odering only on dimenssion can be not predictible. We should order by Y and then x to be sure to have the same order always.

alejoe91 replied:
are you sure it's not predictable?

samuelgarcia commented on spikeinterface/preprocessing/highpass_spatial_filter.py:
```diff
@@ -0,0 +1,290 @@
+            if isinstance(agc_options, str):
+                assert agc_options == "ibl", "agc_options can be 'ibl', a dictionary, or None"
+                # default IBL value is 300 * sampling_interval
+                default_window_length = 300 * sampling_interval 
+                agc_options = {"window_length_s": default_window_length,
```
Here this is strange `window_length_s` make think that it is in second but this is in sample in fact no ?

samuelgarcia replied:
Oups no sorry I missread. this is OK.

samuelgarcia commented on spikeinterface/preprocessing/highpass_spatial_filter.py:
```diff
@@ -0,0 +1,290 @@
+            if isinstance(agc_options, str):
+                assert agc_options == "ibl", "agc_options can be 'ibl', a dictionary, or None"
+                # default IBL value is 300 * sampling_interval
+                default_window_length = 300 * sampling_interval 
```
Maybe we should expose this 300 as a params no ?
What is the units second ?

alejoe91 replied:
you can overwrite the entire dictionary

samuelgarcia commented on spikeinterface/preprocessing/highpass_spatial_filter.py:
```diff
@@ -0,0 +1,290 @@
+
+    traces[:, ~dead_channels] = traces[:, ~dead_channels] / gain[:, ~dead_channels]
+
+    return traces, gain
```
I have the feeling that this gain control will be computed in every chunk.
And so the results will depend on the chunk_size itself.
I am not squre this is a good idea.
Maybe we should compute this once at the begenning no ?

samuelgarcia commented on spikeinterface/preprocessing/highpass_spatial_filter.py:
```diff
@@ -0,0 +1,290 @@
+    else:
+        traces, gain = agc(traces,
+                           window_length=agc_options["window_length_s"],
+                           sampling_interval=agc_options["sampling_interval"])
```
I think the gain control vector should estimated once no ?

samuelgarcia commented on spikeinterface/preprocessing/highpass_spatial_filter.py:
```diff
@@ -0,0 +1,290 @@
+        else:
+            traces = traces.copy()
+
+        traces = kfilt(traces,
+                       self.n_channel_pad,
+                       self.n_channel_taper,
+                       self.agc_options,
+                       self.sos_filter)
```
since this function is mainly a padding and a sosfiltfilt I would not make a call to a function here.
I would move the logic of kfilt just here.
So we can have a better control on what is computed once (in the init) and at evry chunk.

samuelgarcia commented:
Hi joe.
I am reading this.
Thanks again for all that work.
Sorry again for my none forgivable delay.

I have 2 main feedback for this:
  * some tests should be run only locally for developement (like comparing to IBL) and some other should be run the CI
  * I have the feeling that some stuff (taper, gain, ...) should be precomputed in the init instead of every chunk; AM I wrong or is it by design ?

alejoe91 commented:
@samuelgarcia first point is fixed (removed IBL tests on GH).

Also applied suggested changes for phase shift

alejoe91 commented on spikeinterface/preprocessing/highpass_spatial_filter.py:
```diff
@@ -0,0 +1,298 @@
+            order_r = None
+        else:
+            # use stable sort (mergesort) to avoid randomness when non-unique values
+            order_f = np.argsort(locs_mono, kind='mergesort')
```
use lexsort to order x, y (move to core)

alejoe91 commented:
@JoeZiminski with the latest commits we did some further improvements:

- Now we pass plain parameters (`apply_agc` (True/False), `agc_window_length_s`, `highpass_butter_n`, `highpass_butter_wn`)
- Since AGC uses a temporal window, we added a margin when retrieving traces. This makes it chunk-independent
- We removed the `kfilt` function and implemented directly all the steps in the `get_traces`, to be more transparent
- Temporal window and spatial tapering are pre-computed per segment, so they don't need to be recomputed at every get_traces call


@samuelgarcia ready for final review
